design.md

```diff
@@ -0,0 +1,254 @@
+```
+---
+
+### grpc
+
+Package grpc will provide a `JobWorker` type that implements the gRPC server stub.
+
+The `JobWorker` type will be responsible for the following.
+- Ensuring all requests have the necessary permissions to be carried out.
+- Utilize `job.Service` and `cgroups.Service` in coordination to satisfy requests.
+- Validate request inputs.
+
+#### Types
+
+```
+type JobWorker struct {
+  jobService     job.Service
+  cgroupsService cgroups.Service
+}
+
+func (jw JobWorker) Start(ctx context.Context, r *pb.StartRequest) (*pb.StartResponse, error)
+func (jw JobWorker) Stop(ctx context.Context, r *pb.StopRequest) (*pb.StopResponse, error)
+func (jw JobWorker) Status(ctx context.Context, r *pb.StartRequest) (*pb.StatusResponse, error)
+func (jw JobWorker) Output(r *pb.OutputRequest, s pb.JobWokerService_OutputServer) error
+```
+
+## Performance Concerns
+
+If I was going to scale this application, the first component I would revisit is output streaming. In-memory streaming that backed up to disk at some threshold could be considered.
+
+## Testing
+
+I wanted to highlight the areas of the application that are critical to test thoroughly in order to ensure correct function. I will likely not test everything listed as 100% test coverage is not necessary for this challenge.
+
+#### Cgroups
+
+Cgroups have a number of sharp edges and nuances to them. Tests will ensure the following:
+- mounting/unmounting
+- mounting when mount has already been performed
+- setup/cleanup
+- setup when `/cgroup2/jobworker` already exists
+- memory, cpus, and io controls are being applied and are reflected in process stats
+
+#### Authorization
+
+Authorization tests will be fairly simple. They will ensure clients are only able to access jobs they own.
+
+#### Authentication
+
+Authentication tests will ensure the following:
+- client must be using TLS 1.3
+- client certificate must be signed by the CA
+- client that does not authenticate is unable to connect
+
+#### Jobs
+
+Job tests will ensure the following:
+- logging setup/cleanup
+- logging setup when `/var/log/jobworker` already exists
+- job starting and supporting processes result in commands being processed and correct status tracking
+- streaming output results are identical and correct across multiple concurrent clients
+
+## CLI
+
+```
+A simple interface for interacting with a jobworker server.
+
+Usage:
+  jobworker-cli [global flags] command [command flags]
+
+Available Commands:
+  help        Help about any command
+  start       Start job
+  stop        Stop a job
+  status      Fetch a job's status
+  output      Fetch a job's output
+
+Global Flags:
+  --url       URL of jobworker server
+  --cert      client x509 certificate
+  --key       client private key
+  --ca        certificate authority shared by server and client
+
+---
+
+Command:
+ `jobworker-cli start` Start a job. Job details are returned.
+
+Usage:
+  jobworker-cli [global flags] start [flags] command...
```

Comment:
How do I pass arguments to a command?

Reply:
Anything after `jobworker-cli start`'s flags would be considered a command the service should execute.

For instance, `jobworker-cli start --memory 10M ls -la`

Reply:
I've updated the docs to reflect this.